app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/qhybrid/QHybridSupport.java

```diff
@@ -0,0 +1,821 @@
+                // queueWrite(new DownloadFileRequest((short) (256 + ((ListFilesRequest) request).fileCount)));
+            }
+        } else if (request instanceof DownloadFileRequest) {
+            if (((FileRequest) request).completed) {
+                logger.debug("file " + ((DownloadFileRequest) request).fileHandle + " completed: " + ((DownloadFileRequest) request).size);
+                backupFile((DownloadFileRequest) request);
+            }
+        } else if (request instanceof EraseFileRequest) {
+            if (((EraseFileRequest) request).fileHandle > 257) {
+                queueWrite(new DownloadFileRequest((short) (((EraseFileRequest) request).fileHandle - 1)));
+            }
+        }
+        return true;
+    }
+
+    private String arrayToString(byte[] bytes) {
+        if (bytes.length == 0) return "";
+        StringBuilder s = new StringBuilder();
+        final String chars = "0123456789ABCDEF";
+        for (byte b : bytes) {
+            s.append(chars.charAt((b >> 4) & 0xF)).append(chars.charAt(b & 0xF)).append(" ");
+        }
+        return s.substring(0, s.length() - 1) + "\n";
+    }
+
+    private Request resolveAnswer(BluetoothGattCharacteristic characteristic) {
+        byte[] values = characteristic.getValue();
+        if (values[0] != 3) return null;
+        return responseFilters.get(values[1]);
+    }
+
+    private void setHands(short hour, short minute) {
+        queueWrite(new MoveHandsRequest(false, minute, hour, (short) -1));
+    }
+
+    private void vibrate(PlayNotificationRequest.VibrationType vibration) {
+        queueWrite(new PlayNotificationRequest(vibration, -1, -1));
+    }
+
+    private final BroadcastReceiver globalCommandReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            switch (intent.getAction()) {
+                case QHYBRID_ACTION_SET_ACTIVITY_HAND: {
+                    try {
+                        Object extra = intent.getExtras().get("EXTRA_PROGRESS");
+                        float progress = (float) extra;
+                        setActivityHand(progress);
+
+                        queueWrite(
+                                new PlayNotificationRequest(
+                                        PlayNotificationRequest.VibrationType.NO_VIBE,
+                                        -1,
+                                        -1,
+                                        (int) (progress * 180)
+                                )
+                        );
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                        logger.debug("trash extra should be number 0.0-1.0");
+                    }
+                    break;
+                }
+            }
+        }
+    };
+
+    private final BroadcastReceiver commandReceiver = new BroadcastReceiver() {
```

Comment:
Why do you use broadcasts as communication channel instead of the `EventHandler` interface? This is supposed to be extended when new kinds of events/commands are needed.